src/main/java/sparksoniq/jsoniq/runtime/iterator/HybridRuntimeIterator.java

```diff
@@ -101,22 +110,45 @@ public boolean hasNext() {
         }
         if (result == null) {
             currentResultIndex = 0;
-            this._rdd = this.getRDD(_currentDynamicContext);
-            if (SparkSessionManager.LIMIT_COLLECT()) {
-                result = _rdd.take(SparkSessionManager.COLLECT_ITEM_LIMIT);
-                if (result.size() == SparkSessionManager.COLLECT_ITEM_LIMIT) {
-                    if (Main.terminal == null) {
-                        System.out.println("Results have been truncated to:" + SparkSessionManager.COLLECT_ITEM_LIMIT
-                                + " items. This value can be configured with the --result-size parameter at startup.\n");
-                    } else {
-                        Main.terminal.output("\nWarning: Results have been truncated to: " + SparkSessionManager.COLLECT_ITEM_LIMIT
-                                + " items. This value can be configured with the --result-size parameter at startup.\n");
+            if (_isDataFrame) {
+                this._dataFrame = this.getDataFrame(_currentDynamicContext);
+                Row[] rows;
+                if (SparkSessionManager.LIMIT_COLLECT()) {
+                    rows = (Row[]) _dataFrame.take(SparkSessionManager.COLLECT_ITEM_LIMIT);
+                    if (rows.length == SparkSessionManager.COLLECT_ITEM_LIMIT) {
+                        if (Main.terminal == null) {
+                            System.out.println("Results have been truncated to:" + SparkSessionManager.COLLECT_ITEM_LIMIT
+                                    + " items. This value can be configured with the --result-size parameter at startup.\n");
+                        } else {
+                            Main.terminal.output("\nWarning: Results have been truncated to: " + SparkSessionManager.COLLECT_ITEM_LIMIT
+                                    + " items. This value can be configured with the --result-size parameter at startup.\n");
+                        }
                     }
+                } else {
+                    rows = (Row[]) _dataFrame.collect();
+                }
+                for (Row row: rows) {
+                    result.add(JiqsItemParser.getItemFromRow(row, getMetadata()));
```

Comment:
Wondering if this portion of the code is necessary thinking on the medium term.

The DataFrame -> RDD fallback will be implemented in getRDD() soon as part of these efforts (which will compute the RDD from the new overridable getRDDAux() method if no dataframes are available, and otherwise computes the RDD from getDataFrame() with a getItemFromRow map).

-> This will get us the DataFrame -> local fall back for free: the get RDD() call two lines below would take care of the DF conversion already (cascading), with a modular, two-level architecture:

DF    ---- fallback ---->    RDD    ---- fallback ----->     local

What do you think? Anything I am missing?